servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletHttpRequest.java

```diff
@@ -0,0 +1,738 @@
+        return servletContext.getContextPath();
+    }
+
+    @Override
+    public boolean isRequestedSessionIdValid() {
+        return sessionIdSource == SessionTrackingMode.COOKIE ||
+               sessionIdSource == SessionTrackingMode.URL;
+    }
+
+    @Override
+    public boolean isRequestedSessionIdFromCookie() {
+        return sessionIdSource == SessionTrackingMode.COOKIE;
+    }
+
+    @Override
+    public boolean isRequestedSessionIdFromURL() {
+        return isRequestedSessionIdFromUrl();
+    }
+
+    @Override
+    public boolean isRequestedSessionIdFromUrl() {
+        return sessionIdSource == SessionTrackingMode.URL;
+    }
+
+    @Override
+    public String getRequestedSessionId() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    @Nullable
+    public Object getAttribute(String name) {
+        requireNonNull(name, "name");
+        final Object value = getAttributeMap().get(name);
+        return value;
+    }
+
+    @Override
+    public Enumeration<String> getAttributeNames() {
+        return Collections.enumeration(ImmutableSet.copyOf(getAttributeMap().keySet()));
+    }
+
+    @Override
+    public String getCharacterEncoding() {
+        return characterEncoding;
+    }
+
+    @Override
+    public void setCharacterEncoding(String env) throws UnsupportedEncodingException {
+        throw new IllegalStateException("Can't set character encoding after request is initialized");
+    }
+
+    @Override
+    public int getContentLength() {
+        return (int) getContentLengthLong();
+    }
+
+    @Override
+    public long getContentLengthLong() {
+        return Integer.parseInt(getHeader(HttpHeaderNames.CONTENT_LENGTH.toString())
+                                        .replaceAll("[\\[\\]]", ""));
+    }
+
+    @Override
+    @Nullable
+    public String getContentType() {
+        return getHeader(HttpHeaderNames.CONTENT_TYPE.toString());
+    }
+
+    @Override
+    public DefaultServletInputStream getInputStream() throws IOException {
+        if (reader != null) {
+            throw new IllegalStateException("getReader() has already been called for this request");
+        }
+        return inputStream;
+    }
+
+    @Override
+    @Nullable
+    public String getParameter(String name) {
+        requireNonNull(name, "name");
+        final String[] values = getParameterMap().get(name);
+        if (values == null || values.length == 0) {
+            return null;
+        }
+        return values[0];
+    }
+
+    @Override
+    public Enumeration<String> getParameterNames() {
+        return Collections.enumeration(ImmutableSet.copyOf(getParameterMap().keySet()));
+    }
+
+    @Override
+    @Nullable
+    public String[] getParameterValues(String name) {
+        requireNonNull(name, "name");
+        return getParameterMap().get(name);
+    }
+
+    @Override
+    public Map<String, String[]> getParameterMap() {
+        return unmodifiableParameterMap;
```

Comment:
We can build this using the `queryParam`. This will be expensive so we might create this one time in the constructor and reuse it.